xen,pod: Zero-check recently populated pages (checklast)

When demand-populating pages due to guest accesses, check recently populated
pages to see if we can reclaim them for the cache.  This should keep the PoD
cache filled when the start-of-day scrubber is going through.

Signed-off-by: George Dunlap <george.dunlap@eu.citrix.com>

diff -r 7d0468a7f9d9 xen/arch/x86/mm/p2m.c
--- a/xen/arch/x86/mm/p2m.c	Wed Jun 08 12:35:56 2011 +0100
+++ b/xen/arch/x86/mm/p2m.c	Wed Jun 08 12:41:54 2011 +0100
@@ -1113,6 +1113,26 @@ p2m_pod_emergency_sweep_super(struct p2m
     p2m->pod.reclaim_super = i ? i - SUPERPAGE_PAGES : 0;
 }
 
+/* When populating a new superpage, look at recently populated superpages
+ * hoping that they've been zeroed.  This will snap up zeroed pages as soon as 
+ * the guest OS is done with them. */
+static void
+p2m_pod_check_last_super(struct p2m_domain *p2m, unsigned long gfn_aligned)
+{
+    unsigned long check_gfn;
+
+    ASSERT(p2m->pod.last_populated_index < POD_HISTORY_MAX);
+
+    check_gfn = p2m->pod.last_populated[p2m->pod.last_populated_index];
+
+    p2m->pod.last_populated[p2m->pod.last_populated_index] = gfn_aligned;
+
+    p2m->pod.last_populated_index = ( p2m->pod.last_populated_index + 1 ) % POD_HISTORY_MAX;
+
+    p2m->pod.reclaim_super += p2m_pod_zero_check_superpage(p2m, check_gfn);
+}
+
+
 #define POD_SWEEP_STRIDE  16
 static void
 p2m_pod_emergency_sweep(struct p2m_domain *p2m)
@@ -1260,6 +1280,12 @@ p2m_pod_demand_populate(struct p2m_domai
         __trace_var(TRC_MEM_POD_POPULATE, 0, sizeof(t), &t);
     }
 
+    /* Check the last guest demand-populate */
+    if ( p2m->pod.entry_count > p2m->pod.count 
+         && order == 9
+         && q == p2m_guest )
+        p2m_pod_check_last_super(p2m, gfn_aligned);
+
     return 0;
 out_of_memory:
     spin_unlock(&d->page_alloc_lock);
diff -r 7d0468a7f9d9 xen/include/asm-x86/p2m.h
--- a/xen/include/asm-x86/p2m.h	Wed Jun 08 12:35:56 2011 +0100
+++ b/xen/include/asm-x86/p2m.h	Wed Jun 08 12:41:54 2011 +0100
@@ -187,6 +187,7 @@ typedef enum {
 /* Populate-on-demand */
 #define POPULATE_ON_DEMAND_MFN  (1<<9)
 #define POD_PAGE_ORDER 9
+#define POD_HISTORY_MAX 4
 
 #define PAGING_MFN  INVALID_MFN
 
@@ -258,7 +259,9 @@ struct p2m_domain {
         unsigned         reclaim_super; /* Last gpfn of a scan */
         unsigned         reclaim_single; /* Last gpfn of a scan */
         unsigned         max_guest;    /* gpfn of max guest demand-populate */
-    } pod;
+        unsigned         last_populated[POD_HISTORY_MAX]; /* gpfn of last guest page demand-populated */
+        int              last_populated_index;
+  } pod;
 
     struct {
         long entries[3]; 
